### PR TITLE
graph break on intermediate leaves that require grad

### PR DESCRIPTION
**Pull request by bdhirsh**
fixes https://github.com/pytorch/pytorch/issues/90552. This is a simpler fix that just detects the situation where AOTAutograd can't create a proper backward graph for the situation and graph breaks. This was technically a silent correctness issue before.

This PR tries to always graph break when we see a factory function that returns a tensor requiring grad. I check this by seeing if the op returned a `TensorVariable` in dynamo, and if one of the input arguments was a `requires_grad=True` kwarg. I think this is high-fidelity enough, and I'm also hoping that this is uncommon enough that a graph break is reasonable here.

The fix to avoid the graph break in user land is also pretty easy - just instantiate your tensor outside of the compiled region and plumb it in.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113277

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -934,6 +934,21 @@ def f(x, y):
         self.assertEqual(out_ref, out_test)
         self.assertEqual(y_ref, y_test)
 
+    # https://github.com/pytorch/pytorch/issues/90552
+    def test_intermediate_leaf_requires_grad(self):
```
Comment (eellison) [inline test/dynamo/test_repros.py:938]: Can you also comment out all the "aot autograd issue" skips for the nondeterministic tests: https://github.com/pytorch/pytorch/blob/main/test/test_torch.py#L1367 ?

**Review by eellison — APPROVED** [review]
looks good - please remove the fixed skips in test_torch

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (bdhirsh) [reply test/dynamo/test_repros.py:938]: will do

**Comment by bdhirsh** [comment]
@pytorch bot merge

**Pull request closed without merging**